File: client/python/angzarr_client/process_manager.py
"""Base ProcessManager class for stateful cross-domain orchestration.

Process managers coordinate long-running workflows across multiple aggregates.
They maintain their own event-sourced state (keyed by correlation_id) and
react to events from multiple domains.

Two-phase protocol support:
    1. Prepare: Declare additional destinations needed (via @prepares)
    2. Handle: Produce commands and process events (via @reacts_to)

Example usage:
    from angzarr_client import ProcessManager, prepares, reacts_to

    @dataclass
    class OrderWorkflowState:
        order_id: str = ""
        inventory_reserved: bool = False
        payment_received: bool = False

    class OrderWorkflowPM(ProcessManager[OrderWorkflowState]):
        name = "order-workflow"

        def _create_empty_state(self) -> OrderWorkflowState:
            return OrderWorkflowState()

        def _apply_event(self, state, event_any):
            if event_any.type_url.endswith("OrderCreated"):
                ...

        @prepares(OrderCreated)
        def prepare_order(self, event: OrderCreated) -> list[Cover]:
            return [Cover(domain="inventory", root=...)]

        @reacts_to(OrderCreated, input_domain="order")
        def on_order_created(
            self, event: OrderCreated, destinations: list[EventBook]
        ) -> ReserveInventory:
            return ReserveInventory(...)

        @reacts_to(InventoryReserved, input_domain="inventory")
        def on_inventory_reserved(self, event: InventoryReserved) -> ProcessPayment:
            return ProcessPayment(...)
"""

from __future__ import annotations

import inspect
from abc import ABC, abstractmethod
from typing import Generic, TypeVar

from google.protobuf.any_pb2 import Any

from .proto.angzarr import aggregate_pb2 as aggregate
from .proto.angzarr import saga_pb2 as saga
from .proto.angzarr import types_pb2 as types
from .compensation import RejectionHandlerResponse
from .router import _pack_any, prepares, reacts_to, rejected

# Re-export decorators
__all__ = ["ProcessManager", "prepares", "reacts_to", "rejected"]

StateT = TypeVar("StateT")


class ProcessManager(Generic[StateT], ABC):
    """Base class for stateful process managers.

    Process managers are event-sourced aggregates that:
    - React to events from multiple domains
    - Maintain state keyed by correlation_id
    - Produce commands to coordinate workflows
    - Handle rejection/compensation via @rejected handlers
    - Support two-phase protocol with @prepares for destination declaration

    Subclasses must:
    - Set `name` class attribute
    - Implement `_create_empty_state() -> StateT`
    - Implement `_apply_event(state: StateT, event_any: Any) -> None`
    - Decorate event handlers with `@reacts_to(EventType, input_domain="...")`
    - Optionally decorate prepare handlers with `@prepares(EventType)`
    - Optionally decorate rejection handlers with `@rejected(domain, command)`

    Usage:
        class OrderWorkflowPM(ProcessManager[OrderWorkflowState]):
            name = "order-workflow"

            def _create_empty_state(self) -> OrderWorkflowState:
                return OrderWorkflowState()

            def _apply_event(self, state, event_any):
                ...

            @prepares(OrderCreated)
            def prepare_order(self, event: OrderCreated) -> list[Cover]:
                return [Cover(domain="inventory", root=...)]

            @reacts_to(OrderCreated, input_domain="order")
            def on_order_created(
                self, event: OrderCreated, destinations: list[EventBook]
            ) -> ReserveInventory:
                ...

            @rejected(domain="inventory", command="ReserveInventory")
            def handle_reserve_rejected(self, revoke_cmd) -> WorkflowFailed:
                return WorkflowFailed(reason=revoke_cmd.rejection_reason)
    """

    name: str
    _dispatch_table: dict[
        str, tuple[str, type, str, str]
    ] = {}  # suffix -> (method, type, input_domain, output_domain)
    _prepare_table: dict[str, tuple[str, type]] = {}  # suffix -> (method, type)
    _input_domains: dict[str, list[str]] = {}  # domain -> [event types]
    _rejection_table: dict[str, str] = {}  # "domain/command" -> method_name

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        # Skip validation for abstract intermediate classes
        if inspect.isabstract(cls):
            return

        # Validate required class attributes
        if not getattr(cls, "name", None):
            raise TypeError(f"{cls.__name__} must define 'name' class attribute")

        cls._dispatch_table = {}
        cls._prepare_table = {}
        cls._input_domains = {}
        cls._rejection_table = {}
        cls._build_dispatch_table()
        cls._build_prepare_table()
        cls._build_rejection_table()

    @classmethod
    def _build_dispatch_table(cls):
        """Scan for @reacts_to methods and build dispatch table."""
        for attr_name in dir(cls):
            attr = getattr(cls, attr_name, None)
            if callable(attr) and getattr(attr, "_is_handler", False):
                event_type = attr._event_type
                input_domain = attr._input_domain
                output_domain = attr._output_domain
                suffix = event_type.__name__

                if suffix in cls._dispatch_table:
                    raise TypeError(f"{cls.__name__}: duplicate handler for {suffix}")

                cls._dispatch_table[suffix] = (
                    attr_name,
                    event_type,
                    input_domain,
                    output_domain,
                )

                # Track input domains for descriptor
                if input_domain:
                    if input_domain not in cls._input_domains:
                        cls._input_domains[input_domain] = []
                    cls._input_domains[input_domain].append(suffix)

    @classmethod
    def _build_prepare_table(cls):
        """Scan for @prepares methods and build prepare table."""
        for attr_name in dir(cls):
            attr = getattr(cls, attr_name, None)
            if callable(attr) and getattr(attr, "_is_prepare_handler", False):
                event_type = attr._event_type
                suffix = event_type.__name__
                if suffix in cls._prepare_table:
                    raise TypeError(
                        f"{cls.__name__}: duplicate prepare handler for {suffix}"
                    )
                cls._prepare_table[suffix] = (attr_name, event_type)

    @classmethod
    def _build_rejection_table(cls):
        """Scan for @rejected methods and build rejection dispatch table."""
        for attr_name in dir(cls):
            attr = getattr(cls, attr_name, None)
            if callable(attr) and getattr(attr, "_is_rejection_handler", False):
                domain = attr._rejection_domain
                command = attr._rejection_command
                key = f"{domain}/{command}"
                if key in cls._rejection_table:
                    raise TypeError(
                        f"{cls.__name__}: duplicate rejection handler for {key}"
                    )
                cls._rejection_table[key] = attr_name

    def __init__(self, process_state: types.EventBook = None):
        """Initialize process manager with optional prior state.

        Args:
            process_state: Existing event book for this correlation_id.
        """
        if process_state is None:
            process_state = types.EventBook()
        self._event_book = process_state
        self._state: StateT = None
        self._new_events: list[Any] = []

    def _get_state(self) -> StateT:
        """Get current state, rebuilding from events if needed."""
        if self._state is None:
            self._state = self._rebuild()
        return self._state

    def _rebuild(self) -> StateT:
        """Rebuild state from event book."""
        state = self._create_empty_state()
        for page in self._event_book.pages:
            if page.event:
                self._apply_event(state, page.event)
        return state

    def prepare(self, event_any: Any) -> list[types.Cover]:
        """Prepare destinations for an event.

        Dispatches to @prepares decorated method if one exists.

        Args:
            event_any: Packed event as google.protobuf.Any

        Returns:
            List of Covers identifying destination aggregates.
        """
        type_url = event_any.type_url

        for suffix, (method_name, event_type) in self._prepare_table.items():
            if type_url.endswith(suffix):
                # Unpack event
                event = event_type()
                event_any.Unpack(event)

                # Call prepare handler
                result = getattr(self, method_name)(event)
                return result if result else []

        return []

    def dispatch(
        self,
        event_any: Any,
        root: bytes = None,
        correlation_id: str = "",
        destinations: list[types.EventBook] = None,
    ) -> list[types.CommandBook]:
        """Dispatch event to matching handler.

        Args:
            event_any: Packed event as google.protobuf.Any
            root: Source aggregate root
            correlation_id: Correlation ID for the workflow
            destinations: Optional list of destination EventBooks

        Returns:
            List of CommandBooks to send.
        """
        type_url = event_any.type_url

        for suffix, (
            method_name,
            event_type,
            _,
            output_domain,
        ) in self._dispatch_table.items():
            if type_url.endswith(suffix):
                # Unpack event
                event = event_type()
                event_any.Unpack(event)

                # Check if handler accepts destinations parameter
                method = getattr(self, method_name)
                sig = inspect.signature(method)
                params = list(sig.parameters.keys())

                # Call handler with or without destinations
                if "destinations" in params:
                    result = method(event, destinations=destinations or [])
                else:
                    result = method(event)

                # Pack result into CommandBooks
                return self._pack_commands(result, output_domain, root, correlation_id)

        return []

    def _pack_commands(
        self,
        result,
        output_domain: str = None,
        root: bytes = None,
        correlation_id: str = "",
    ) -> list[types.CommandBook]:
        """Pack command(s) into CommandBooks."""
        if result is None:
            return []

        # Handle pre-packed CommandBooks (advanced usage)
        if isinstance(result, types.CommandBook):
            return [result]
        if (
            isinstance(result, list)
            and result
            and isinstance(result[0], types.CommandBook)
        ):
            return result

        commands = result if isinstance(result, tuple) else (result,)
        books = []

        for cmd in commands:
            cmd_any = _pack_any(cmd)
            # Use handler's output_domain if specified
            domain = output_domain or ""
            cover = types.Cover(
                domain=domain,
                correlation_id=correlation_id,
            )
            if root:
                cover.root.value = root

            book = types.CommandBook(
                cover=cover,
                pages=[types.CommandPage(command=cmd_any)],
            )
            books.append(book)

        return books

    def _apply_and_record(self, event) -> None:
        """Pack event, apply to cached state, record for persistence."""
        event_any = _pack_any(event)

        # Apply directly to cached state
        if self._state is not None:
            self._apply_event(self._state, event_any)

        self._new_events.append(event_any)

    def process_events(self) -> types.EventBook:
        """Return new process events for persistence."""
        pages = [types.EventPage(event=e) for e in self._new_events]
        return types.EventBook(pages=pages)

    @classmethod
    def prepare_destinations(
        cls,
        trigger: types.EventBook,
        process_state: types.EventBook,
    ) -> list[types.Cover]:
        """Phase 1: Declare additional destinations needed.

        Args:
            trigger: The triggering event book.
            process_state: Current process manager state.

        Returns:
            List of Covers identifying destination aggregates to fetch.
        """
        pm = cls(process_state)
        destinations: list[types.Cover] = []

        for page in trigger.pages:
            if page.HasField("event"):
                destinations.extend(pm.prepare(page.event))

        return destinations

    @classmethod
    def handle(
        cls,
        trigger: types.EventBook,
        process_state: types.EventBook,
        destinations: list[types.EventBook] = None,
    ) -> tuple[list[types.CommandBook], types.EventBook]:
        """Phase 2: Handle a trigger event with current process state.

        Args:
            trigger: The triggering event book.
            process_state: Current process manager state.
            destinations: Additional destination states (from prepare phase).

        Returns:
            Tuple of (commands to send, new process events).
        """
        pm = cls(process_state)
        root = trigger.cover.root.value if trigger.HasField("cover") else None
        correlation_id = (
            trigger.cover.correlation_id if trigger.HasField("cover") else ""
        )

        commands = []
        for page in trigger.pages:
            if page.HasField("event"):
                commands.extend(
                    pm.dispatch(page.event, root, correlation_id, destinations)
                )

        return commands, pm.process_events()

    def handle_revocation(
        self,
        notification: types.Notification,
    ) -> RejectionHandlerResponse:
        """Handle a rejection notification for commands this PM issued.

        Called when a command produced by this PM is rejected by the target aggregate.
        Dispatches to @rejected decorated methods based on target domain
        and rejected command type.

        If no matching @rejected handler is found, returns empty response
        (framework handles default behavior).

        Args:
            notification: Notification containing RejectionNotification payload.

        Returns:
            RejectionHandlerResponse with optional events (compensation)
            and/or notification (upstream propagation).

        Usage:
            @rejected(domain="inventory", command="ReserveInventory")
            def handle_reserve_rejected(self, notification) -> RejectionHandlerResponse:
                ctx = CompensationContext.from_notification(notification)
                self._apply_and_record(WorkflowFailed(reason=ctx.rejection_reason))
                return RejectionHandlerResponse(events=self.process_events())
        """
        # Unpack rejection details from notification payload
        rejection = types.RejectionNotification()
        if notification.HasField("payload"):
            notification.payload.Unpack(rejection)

        # Extract domain and command type from rejected_command
        domain = ""
        command_suffix = ""

        if rejection.HasField("rejected_command") and rejection.rejected_command.pages:
            rejected_cmd = rejection.rejected_command
            if rejected_cmd.HasField("cover"):
                domain = rejected_cmd.cover.domain
            if rejected_cmd.pages[0].HasField("command"):
                type_url = rejected_cmd.pages[0].command.type_url
                # Extract suffix (e.g., "ReserveInventory" from "type.googleapis.com/.../ReserveInventory")
                command_suffix = (
                    type_url.rsplit("/", 1)[-1] if "/" in type_url else type_url
                )

        # Dispatch to @rejected handler if found (use suffix matching like regular dispatch)
        for key, method_name in self._rejection_table.items():
            expected_domain, expected_command = key.split("/", 1)
            if domain == expected_domain and command_suffix.endswith(expected_command):
                # Ensure state is built before calling handler
                _ = self._get_state()
                # Call the handler
                result = getattr(self, method_name)(notification)

                # Handler may return RejectionHandlerResponse directly
                if isinstance(result, RejectionHandlerResponse):
                    return result

                # Handler may return None or a single event
                process_events = self.process_events()
                return RejectionHandlerResponse(
                    events=process_events if process_events.pages else None,
                )

        # Default: no handler, return empty response
        return RejectionHandlerResponse()

    @property
    def state(self) -> StateT:
        """Get current state (convenience property for _get_state)."""
        return self._get_state()

    @abstractmethod
    def _create_empty_state(self) -> StateT:
        """Create an empty state instance. Must be implemented by subclasses."""
        ...

    @abstractmethod
    def _apply_event(self, state: StateT, event_any: Any) -> None:
        """Apply a single event to state. Must be implemented by subclasses."""
        ...

File: examples/csharp/Angzarr.Client/ProcessManager.cs
using System.Reflection;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Type = System.Type;

namespace Angzarr.Client;

/// <summary>
/// Base class for stateful process managers using the OO pattern.
///
/// Subclasses must:
/// - Override Name property
/// - Override CreateEmptyState() and ApplyEvent()
/// - Decorate event handlers with [ReactsTo(typeof(EventType), InputDomain = "...")]
/// - Optionally decorate prepare handlers with [Prepares(typeof(EventType))]
/// - Optionally decorate rejection handlers with [Rejected("domain", "command")]
/// </summary>
public abstract class ProcessManager<TState> where TState : class
{
    private Angzarr.EventBook _eventBook;
    private TState? _state;
    private readonly List<Any> _newEvents = new();

    // Dispatch tables built via reflection on first use
    private static readonly Dictionary<Type, Dictionary<string, (MethodInfo Method, Type EventType, string? InputDomain, string? OutputDomain)>> _dispatchTables = new();
    private static readonly Dictionary<Type, Dictionary<string, (MethodInfo Method, Type EventType)>> _prepareTables = new();
    private static readonly Dictionary<Type, Dictionary<string, MethodInfo>> _rejectionTables = new();
    private static readonly Dictionary<Type, Dictionary<string, List<string>>> _inputDomains = new();

    /// <summary>
    /// The name of this process manager.
    /// </summary>
    public abstract string Name { get; }

    /// <summary>
    /// Create an empty state instance.
    /// </summary>
    protected abstract TState CreateEmptyState();

    /// <summary>
    /// Apply a single event to state.
    /// </summary>
    protected abstract void ApplyEvent(TState state, Any eventAny);

    protected ProcessManager(Angzarr.EventBook? processState = null)
    {
        _eventBook = processState ?? new Angzarr.EventBook();
        EnsureDispatchTablesBuilt();
    }

    private void EnsureDispatchTablesBuilt()
    {
        var type = GetType();
        if (_dispatchTables.ContainsKey(type)) return;

        lock (_dispatchTables)
        {
            if (_dispatchTables.ContainsKey(type)) return;

            var dispatch = new Dictionary<string, (MethodInfo, Type, string?, string?)>();
            var domains = new Dictionary<string, List<string>>();

            foreach (var method in type.GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
            {
                var attr = method.GetCustomAttribute<ReactsToAttribute>();
                if (attr != null)
                {
                    var suffix = attr.EventType.Name;
                    dispatch[suffix] = (method, attr.EventType, attr.InputDomain, attr.OutputDomain);

                    // Track input domains
                    if (!string.IsNullOrEmpty(attr.InputDomain))
                    {
                        if (!domains.TryGetValue(attr.InputDomain, out var types))
                        {
                            types = new List<string>();
                            domains[attr.InputDomain] = types;
                        }
                        types.Add(suffix);
                    }
                }
            }
            _dispatchTables[type] = dispatch;
            _inputDomains[type] = domains;

            // Build prepare handler dispatch table
            var prepares = new Dictionary<string, (MethodInfo, Type)>();
            foreach (var method in type.GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
            {
                var attr = method.GetCustomAttribute<PreparesAttribute>();
                if (attr != null)
                {
                    var suffix = attr.EventType.Name;
                    prepares[suffix] = (method, attr.EventType);
                }
            }
            _prepareTables[type] = prepares;

            // Build rejection handler dispatch table
            var rejections = new Dictionary<string, MethodInfo>();
            foreach (var method in type.GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
            {
                var attr = method.GetCustomAttribute<RejectedAttribute>();
                if (attr != null)
                {
                    var key = $"{attr.Domain}/{attr.Command}";
                    rejections[key] = method;
                }
            }
            _rejectionTables[type] = rejections;
        }
    }

    /// <summary>
    /// Phase 1: Declare additional destinations needed.
    /// </summary>
    public static List<Angzarr.Cover> PrepareDestinations<T>(
        Angzarr.EventBook trigger,
        Angzarr.EventBook processState) where T : ProcessManager<TState>, new()
    {
        var pm = (T)Activator.CreateInstance(typeof(T), processState)!;
        var destinations = new List<Angzarr.Cover>();

        foreach (var page in trigger.Pages)
        {
            if (page.Event != null)
                destinations.AddRange(pm.Prepare(page.Event));
        }

        return destinations;
    }

    /// <summary>
    /// Phase 2: Handle a trigger event with current process state.
    /// </summary>
    public static (List<Angzarr.CommandBook> Commands, Angzarr.EventBook ProcessEvents) Handle<T>(
        Angzarr.EventBook trigger,
        Angzarr.EventBook processState,
        List<Angzarr.EventBook>? destinations = null) where T : ProcessManager<TState>, new()
    {
        var pm = (T)Activator.CreateInstance(typeof(T), processState)!;
        var root = trigger.Cover?.Root?.Value.ToByteArray();
        var correlationId = trigger.Cover?.CorrelationId ?? "";

        var commands = new List<Angzarr.CommandBook>();
        foreach (var page in trigger.Pages)
        {
            if (page.Event != null)
                commands.AddRange(pm.Dispatch(page.Event, root, correlationId, destinations));
        }

        return (commands, pm.ProcessEvents());
    }

    /// <summary>
    /// Prepare destinations for a single event.
    /// </summary>
    public List<Angzarr.Cover> Prepare(Any eventAny)
    {
        var prepareTable = _prepareTables[GetType()];
        foreach (var (suffix, (method, eventType)) in prepareTable)
        {
            if (eventAny.TypeUrl.EndsWith(suffix))
            {
                var unpackMethod = typeof(Any).GetMethod("Unpack")!.MakeGenericMethod(eventType);
                var evt = unpackMethod.Invoke(eventAny, null);
                var result = method.Invoke(this, new[] { evt });
                return result as List<Angzarr.Cover> ?? new List<Angzarr.Cover>();
            }
        }
        return new List<Angzarr.Cover>();
    }

    /// <summary>
    /// Dispatch a single event to the matching handler.
    /// </summary>
    public List<Angzarr.CommandBook> Dispatch(
        Any eventAny,
        byte[]? root = null,
        string correlationId = "",
        List<Angzarr.EventBook>? destinations = null)
    {
        var dispatchTable = _dispatchTables[GetType()];
        foreach (var (suffix, (method, eventType, _, outputDomain)) in dispatchTable)
        {
            if (eventAny.TypeUrl.EndsWith(suffix))
            {
                var unpackMethod = typeof(Any).GetMethod("Unpack")!.MakeGenericMethod(eventType);
                var evt = unpackMethod.Invoke(eventAny, null);

                // Check if method accepts destinations parameter
                var parameters = method.GetParameters();
                object? result;
                if (parameters.Any(p => p.Name == "destinations"))
                {
                    result = method.Invoke(this, new object?[] { evt, destinations ?? new List<Angzarr.EventBook>() });
                }
                else
                {
                    result = method.Invoke(this, new[] { evt });
                }

                return PackCommands(result, outputDomain, root, correlationId);
            }
        }
        return new List<Angzarr.CommandBook>();
    }

    /// <summary>
    /// Handle rejection notification.
    /// </summary>
    public (Angzarr.EventBook? Events, Angzarr.RevocationResponse Revocation) HandleRevocation(
        Angzarr.Notification notification)
    {
        var rejection = notification.Payload?.Unpack<Angzarr.RejectionNotification>();
        var domain = "";
        var commandSuffix = "";

        if (rejection?.RejectedCommand?.Pages.Count > 0)
        {
            var rejectedCmd = rejection.RejectedCommand;
            domain = rejectedCmd.Cover?.Domain ?? "";
            var cmdTypeUrl = rejectedCmd.Pages[0].Command?.TypeUrl ?? "";
            commandSuffix = Helpers.TypeNameFromUrl(cmdTypeUrl);
        }

        var rejectionTable = _rejectionTables[GetType()];
        foreach (var (key, method) in rejectionTable)
        {
            var parts = key.Split('/');
            if (parts[0] == domain && commandSuffix.EndsWith(parts[1]))
            {
                _ = State; // Ensure state is built
                var result = method.Invoke(this, new object[] { notification });
                HandleResult(result);
                return (ProcessEvents(), new Angzarr.RevocationResponse
                {
                    EmitSystemRevocation = false,
                    Reason = $"ProcessManager {Name} handled rejection for {key}"
                });
            }
        }

        return (null, new Angzarr.RevocationResponse
        {
            EmitSystemRevocation = true,
            Reason = $"ProcessManager {Name} has no custom compensation for {domain}/{commandSuffix}"
        });
    }

    /// <summary>
    /// Get the current state.
    /// </summary>
    public TState State => GetState();

    /// <summary>
    /// Return new process events for persistence.
    /// </summary>
    public Angzarr.EventBook ProcessEvents()
    {
        var book = new Angzarr.EventBook();
        book.Pages.AddRange(_newEvents.Select(e => new Angzarr.EventPage { Event = e }));
        return book;
    }

    /// <summary>
    /// Build a component descriptor for topology discovery.
    /// </summary>
    public Descriptor Descriptor()
    {
        var domains = _inputDomains[GetType()];
        var inputs = domains.Select(kv => new TargetDesc(kv.Key, kv.Value)).ToList();
        return new Descriptor(Name, ComponentTypes.ProcessManager, inputs);
    }

    private TState GetState()
    {
        if (_state == null)
            _state = Rebuild();
        return _state;
    }

    private TState Rebuild()
    {
        var state = CreateEmptyState();
        foreach (var page in _eventBook.Pages)
        {
            if (page.Event != null)
                ApplyEvent(state, page.Event);
        }
        return state;
    }

    /// <summary>
    /// Pack event, apply to cached state, record for persistence.
    /// </summary>
    protected void ApplyAndRecord(IMessage eventMessage)
    {
        var eventAny = Any.Pack(eventMessage, "type.googleapis.com/");

        if (_state != null)
            ApplyEvent(_state, eventAny);

        _newEvents.Add(eventAny);
    }

    private void HandleResult(object? result)
    {
        if (result == null) return;

        if (result is System.Collections.IEnumerable enumerable && result is not IMessage)
        {
            foreach (var item in enumerable)
            {
                if (item is IMessage msg)
                    ApplyAndRecord(msg);
            }
        }
        else if (result is IMessage message)
        {
            ApplyAndRecord(message);
        }
    }

    private List<Angzarr.CommandBook> PackCommands(
        object? result,
        string? outputDomain,
        byte[]? root,
        string correlationId)
    {
        if (result == null) return new List<Angzarr.CommandBook>();

        // Handle pre-packed CommandBooks
        if (result is Angzarr.CommandBook book)
            return new List<Angzarr.CommandBook> { book };
        if (result is List<Angzarr.CommandBook> books)
            return books;

        var commands = new List<IMessage>();
        if (result is System.Collections.IEnumerable enumerable && result is not IMessage)
        {
            foreach (var item in enumerable)
            {
                if (item is IMessage msg)
                    commands.Add(msg);
            }
        }
        else if (result is IMessage message)
        {
            commands.Add(message);
        }

        return commands.Select(cmd => PackCommand(cmd, outputDomain ?? "", root, correlationId)).ToList();
    }

    private Angzarr.CommandBook PackCommand(IMessage cmd, string domain, byte[]? root, string correlationId)
    {
        var cmdAny = Any.Pack(cmd, "type.googleapis.com/");
        var cover = new Angzarr.Cover
        {
            Domain = domain,
            CorrelationId = correlationId
        };
        if (root != null)
            cover.Root = new Angzarr.UUID { Value = Google.Protobuf.ByteString.CopyFrom(root) };

        return new Angzarr.CommandBook
        {
            Cover = cover,
            Pages = { new Angzarr.CommandPage { Command = cmdAny } }
        };
    }
}

/// <summary>
/// Helper utilities for ProcessManager.
/// </summary>
internal static class Helpers
{
    /// <summary>
    /// Extract the type name suffix from a type_url.
    /// </summary>
    public static string TypeNameFromUrl(string typeUrl)
    {
        var lastSlash = typeUrl.LastIndexOf('/');
        return lastSlash >= 0 ? typeUrl.Substring(lastSlash + 1) : typeUrl;
    }
}
